thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/DefaultTimeSeriesCache.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.pinot.thirdeye.common.time.TimeSpec;
+import org.apache.pinot.thirdeye.dataframe.util.DataFrameUtils;
+import org.apache.pinot.thirdeye.dataframe.util.MetricSlice;
+import org.apache.pinot.thirdeye.dataframe.util.TimeSeriesRequestContainer;
+import org.apache.pinot.thirdeye.datalayer.bao.DatasetConfigManager;
+import org.apache.pinot.thirdeye.datalayer.bao.MetricConfigManager;
+import org.apache.pinot.thirdeye.datalayer.dto.DatasetConfigDTO;
+import org.apache.pinot.thirdeye.datasource.MetricFunction;
+import org.apache.pinot.thirdeye.datasource.RelationalThirdEyeResponse;
+import org.apache.pinot.thirdeye.datasource.ThirdEyeRequest;
+import org.apache.pinot.thirdeye.datasource.ThirdEyeResponse;
+import org.apache.pinot.thirdeye.datasource.TimeRangeUtils;
+import org.apache.pinot.thirdeye.datasource.cache.QueryCache;
+import org.apache.pinot.thirdeye.rootcause.impl.MetricEntity;
+import org.apache.pinot.thirdeye.util.ThirdEyeUtils;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class DefaultTimeSeriesCache implements TimeSeriesCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultTimeSeriesCache.class);
+
+  private final MetricConfigManager metricDAO;
+  private final DatasetConfigManager datasetDAO;
+  private final QueryCache queryCache;
+  private CouchbaseCacheDAO cacheDAO = new CouchbaseCacheDAO();
+
+  public DefaultTimeSeriesCache(MetricConfigManager metricDAO, DatasetConfigManager datasetDAO, QueryCache queryCache) {
+    this.metricDAO = metricDAO;
+    this.datasetDAO = datasetDAO;
+    this.queryCache = queryCache;
+  }
+
+  public ThirdEyeResponse fetchTimeSeries(ThirdEyeRequest thirdEyeRequest) throws Exception {
+    //LOG.info("trying to fetch data from cache...");
+
+    if (!CacheConfig.useCentralizedCache()) {
+      return this.queryCache.getQueryResult(thirdEyeRequest);
+    }
+
+    ThirdEyeCacheResponse cacheResponse = cacheDAO.tryFetchExistingTimeSeries(ThirdEyeCacheRequest.from(thirdEyeRequest));
+
+    DateTime sliceStart = thirdEyeRequest.getStartTimeInclusive();
+    DateTime sliceEnd = thirdEyeRequest.getEndTimeExclusive();
+
+    if (cacheResponse.isMissingSlice(sliceStart.getMillis(), sliceEnd.getMillis())) {
+      fetchMissingSlices(cacheResponse);
+    }
+
+    return buildThirdEyeResponseFromCacheResponse(cacheResponse);
+  }
+
+  // fetches missing time-series slices and merges them into the cacheResponse object.
+  private void fetchMissingSlices(ThirdEyeCacheResponse cacheResponse) throws Exception {
+
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+    long metricId = request.getMetricFunctions().get(0).getMetricId();
+    long requestSliceStart = request.getStartTimeInclusive().getMillis();
+    long requestSliceEnd = request.getEndTimeExclusive().getMillis();
+
+    ThirdEyeResponse result;
+    MetricSlice slice;
```

Comment:
Is it possible that the missing slice is in the middle?

Reply:
Yes, it's possible but probably fairly rare. The case where this happens is when the documents for a specific timeseries are about to expire in the cache, and there is a cache fetch during the small slice of time when some of the documents have expired, but others have not. Also, it's difficult to know whether the missing documents in the middle are because there are actually missing documents, or if these are missing data points from the data source itself. Since this case is so complicated and fairly rare, I decided to ignore it and pretend it doesn't happen for now. It can be added, but may require us to refetch the entire timeseries from the data source to ensure data integrity, which we're trying to avoid as much as possible.

Reply:
Can we add a sanity check to make sure there is no whole in the middle.

Reply:
added, it will log any missing timeseries points in the middle of the timeseries pulled from the cache. it will produce noise if the original time-series is missing data points in the middle of the time-series, the only way to fix this is to insert extra documents with "null" as the value, and skip those data points when putting the cached data back together.